File: src/python/xerxes_agent/channels/adapters/dingtalk.py
# Copyright 2025 The EasyDeL/Xerxes Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License")
"""DingTalk (Alibaba) bot adapter."""

from __future__ import annotations

import typing as tp

from .._helpers import WebhookChannel, http_post, parse_json_body
from ..types import ChannelMessage, MessageDirection


class DingTalkChannel(WebhookChannel):
    """DingTalk custom-bot adapter (incoming webhook + access_token send)."""

    name = "dingtalk"

    def __init__(
        self,
        webhook_url: str,
        *,
        http_client: tp.Any = None,
    ) -> None:
        """Bind the adapter to a DingTalk custom-bot webhook.

        Args:
            webhook_url: Full webhook URL including the ``access_token``
                query parameter provided by DingTalk.
            http_client: Optional injected HTTP callable for tests.
        """
        super().__init__()
        self.webhook_url = webhook_url
        self._http = http_client

    def _parse_inbound(self, headers, body):
        """Extract ``text.content`` / ``content``, sender id and conversation id."""
        data = parse_json_body(body)
        if not data:
            return []
        text = (data.get("text") or {}).get("content", "") or data.get("content", "")
        return [
            ChannelMessage(
                text=text,
                channel=self.name,
                channel_user_id=data.get("senderId", "") or data.get("senderStaffId", ""),
                room_id=data.get("conversationId", ""),
                platform_message_id=str(data.get("msgId", "")),
                direction=MessageDirection.INBOUND,
                metadata={"sender_nick": data.get("senderNick", "")},
            )
        ]

    async def _send_outbound(self, message):
        """POST a ``msgtype=text`` envelope back to the bot webhook URL."""
        body = {"msgtype": "text", "text": {"content": message.text}}
        http_post(self.webhook_url, json_body=body, http_client=self._http)

File: xerxes/src/channels/dingtalk.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import { postJson, type ChannelFetch } from './http.js'
import {
  recordValue,
  requiredOption,
  stringValue,
  throwOnProviderErrorCode,
  type RelayChannelTransport,
} from './relay.js'
import { createChannelMessage, MessageDirection, type ChannelMessage } from './types.js'
import { parseJsonBody, WebhookChannel, type WebhookHeaders } from './webhooks.js'

/** DingTalk's signed/stream transports stay upstream; this uses webhook relay only. */
export const DINGTALK_TRANSPORT: RelayChannelTransport = {
  inbound: 'webhook-relay',
  outbound: 'incoming-webhook',
  unsupported: ['DingTalk stream-mode connections', 'incoming webhook signature verification'],
}

export interface DingTalkChannelOptions {
  readonly fetchImplementation?: ChannelFetch
  /** Conversation-scoped incoming webhook URL, including any access token. */
  readonly webhookUrl: string
}

/**
 * DingTalk outgoing-webhook relay and incoming-webhook sender.
 *
 * The configured URL determines the outbound conversation. Signature checks
 * for public inbound endpoints remain the responsibility of the HTTP edge.
 */
export class DingTalkChannel extends WebhookChannel {
  readonly name = 'dingtalk'
  readonly transport = DINGTALK_TRANSPORT

  private readonly fetchImplementation: ChannelFetch | undefined
  private readonly webhookUrl: string

  constructor(options: DingTalkChannelOptions) {
    super()
    this.fetchImplementation = options.fetchImplementation
    this.webhookUrl = requiredOption(options.webhookUrl, 'DingTalk webhookUrl')
  }

  protected parseInbound(
    _headers: WebhookHeaders,
    body: Uint8Array,
  ): readonly ChannelMessage[] {
    const payload = parseJsonBody(body)
    if (!Object.keys(payload).length) {
      return []
    }
    const text = stringValue(recordValue(payload.text).content) || stringValue(payload.content)
    if (!text) {
      return []
    }
    return [createChannelMessage({
      channel: this.name,
      direction: MessageDirection.INBOUND,
      text,
      channelUserId: stringValue(payload.senderId) || stringValue(payload.senderStaffId),
      roomId: stringValue(payload.conversationId),
      platformMessageId: stringValue(payload.msgId),
      metadata: { sender_nick: stringValue(payload.senderNick) },
    })]
  }

  protected async sendOutbound(message: ChannelMessage): Promise<void> {
    const response = await postJson(this.webhookUrl, {
      body: { msgtype: 'text', text: { content: message.text } },
      ...(this.fetchImplementation ? { fetchImplementation: this.fetchImplementation } : {}),
    })
    throwOnProviderErrorCode(response, 'DingTalk', 'errcode')
  }
}
